### PR TITLE
[New Concept Exercise on `lambda`]: Add Secrets draft

### LINKED ISSUE #2357: OLD anonymous-functions (lambdas) concept exercise
This issue describes how to implement the `anonymous-functions` (lambdas) concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/creation/use of `lambda` or `anonymous functions`  in python. 

## Learning objectives

- Understand what an `anonymous function` is, and how to create one
   - The syntax of creating a  `lambda` 
   - Using different `function argument` flavors with `lambda`
- Understand the differences between `lambdas` and Pythons "regular" `functions`
- Understand what problems are solved by using a `lambda`
- The pitfalls of `lambdas`, and when to avoid them
- Using `lambdas` as `key functions` in other situations such as `sort()` ,  `sorted()`, `min()`,  and `max()`
- Applying arguments to a `lambda` via IIFE (_immediately invoked function expression_)
- Anti-patterns when using `lambdas` 

  

## Out of scope

- `comprehensions`
- `comprehensions` in `lambdas`
- using a `decorator` on a `lambda`
- `functools` (_this will get its own exercise_)
- `generators`
- `map()`, `filter()`, and `reduce()` (_these will get their own exercise_)
- using an `assignment expression` or "walrus" operator (`:=`) in a `lambda`

  

## Concepts

- `anonymous-functions`
- `lambdas`
- `functions`, 
- `higher-order functions`
- `functions as arguments`
- `functions as returns`
- `nested funcitons`

  

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `booleans`
- `comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
- `higher-order functions`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

  

## Resources to refer to

- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Python Docs Tutorial: Lambda Expressions](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions)
- [Functions as Objects in Python](https://medium.com/python-pandemonium/function-as-objects-in-python-d5215e6d1b0d)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [Learn by Example:  Python Lambda Function](https://www.learnbyexample.org/python-lambda-function/)
- [Real Python: How to Use Python Lambda Fuctions](https://realpython.com/python-lambda/)
- [Trey Hunner: Overusing Lambda expressions in Python](https://treyhunner.com/2018/09/stop-writing-lambda-expressions/)
  

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  * You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  *  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  *  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

<br>

## Concept Description

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts) & [concept exercises](https://exercism.org/docs/building/tracks/concept-exercises)

* ### Concept `about.md`

  **Concept file/issue**: [anonymous-functions directory with stubbed files](https://github.com/exercism/python/tree/main/concepts/anonymous-functions) -- **Content is TBD** and should be completed as part of this exercise creation.  `Anonymous-functions` concept write-ups and associated files can be included in the PR for this issue, or as a separate PR linked to this issue.

    For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)
  
     - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

<br>  

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

<br>

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the [root of this `Python` repo](https://github.com/exercism/python/blob/main/config.json).

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

<br>

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.

- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

  - All asserts should contain a "user friendly" failure message (_these will display on the website_). 
  - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
  - We also use `unittest.subtest` to parameterize test input where/when needed. 
    Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

### PR DESCRIPTION
**Pull request by Steffan153**
Closes #2357

My checklist:
- [x] Write introduction
- [x] Write instructions
- [x] Write tests
- [x] Write skeleton file
- [x] Write exemplar file
- [x] Write hints
- [x] Write config.json
- [x] Write design.md
- [x] Write concept introduction.md
- [ ] Write concept about.md
- [x] Write concept config.json
- [x] Add something about using `sorted`

### CODE REVIEW

File: exercises/concept/secrets/.meta/exemplar.py
Comment (IsaacG) [inline exercises/concept/secrets/.meta/exemplar.py:5]: This looks like subtraction

**Review by IsaacG — CHANGES_REQUESTED** [review]

File: exercises/concept/secrets/.meta/exemplar.py
Comment (Steffan153) [reply exercises/concept/secrets/.meta/exemplar.py:5]: Smh how on earth did THAT happen?!?

File: exercises/concept/secrets/.docs/instructions.md
Comment (BethanyG) [inline exercises/concept/secrets/.docs/instructions.md:16]:
```suggestion
```python
>>> adder = secret_add(2)
>>> adder(2)
4
```

```

Comment (BethanyG) [reply exercises/concept/secrets/.docs/instructions.md:16]: For better (or worse) we've decided to go the "python terminal" route with code examples.  I'll just leave the one comment - it's a small thing to clean up later.  😄

File: exercises/concept/secrets/.docs/instructions.md
Comment (BethanyG) [inline exercises/concept/secrets/.docs/instructions.md]: The remainder of these code examples should also be changed per the code block above.

Comment (Steffan153) [reply exercises/concept/secrets/.docs/instructions.md]: You're late. I already did. :)

File: exercises/concept/secrets/.docs/introduction.md
Comment (BethanyG) [inline exercises/concept/secrets/.docs/introduction.md:3]: Maybe something here about them being "nameless" or anonymous?

File: exercises/concept/secrets/.docs/introduction.md
Comment (BethanyG) [inline exercises/concept/secrets/.docs/introduction.md:5]: Assigning a `lambda` to a variable is a [Python Anti-pattern](https://docs.quantifiedcode.com/python-anti-patterns/correctness/assigning_a_lambda_to_a_variable.html)

File: exercises/concept/secrets/.docs/introduction.md
Comment (BethanyG) [inline exercises/concept/secrets/.docs/introduction.md:9]:
```suggestion
In short, a `lambda expression` starts with the keyword `lambda`, followed by parameters (_separated by a comma, as you would dow with a `def`-defined function_), a colon, and a `return` value.
```

**Comment by Steffan153** [comment]
I'm not sure what to put in `about.md`. I think somtehing about `sorted` would be good, but I'm not sure how to relate it with secrets. Maybe the secret has somtehing do do iwth the `key`?

**Comment by Steffan153** [comment]
I rigged up something for the sort.

File: config.json
Comment (BethanyG) [inline config.json:2287]: This is a concept exercise, so needs to follow the [concept exercise](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata) specifications for `config.json`.

File: config.json
Comment (BethanyG) [inline config.json:2296]:
```suggestion
          "higher-orderf-unctions",
```

Comment (BethanyG) [reply config.json:2296]: This is causing the CI lint to fail, so needs to be completely kebab-case.

File: config.json
Comment (BethanyG) [inline config.json:2307]:
```suggestion
        "status": "wip"
```

File: exercises/concept/secrets/secrets.py
Comment (BethanyG) [inline exercises/concept/secrets/secrets.py:1]: Docstrings should be added for all functions here.  Please see [this exercise](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/conditionals.py) for an example of the format.

File: exercises/concept/secrets/.meta/exemplar.py
Comment (BethanyG) [inline exercises/concept/secrets/.meta/exemplar.py:1]: Docstrings should be added for all functions here.  Please see [this exemplar.py file](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example of the format.

File: exercises/concept/secrets/secrets_test.py
Comment (BethanyG) [inline exercises/concept/secrets/secrets_test.py:53]:
```suggestion
        self.assertEqual(secret_combine(secret_max([3, 8]), sum)([4, 5]), 11)
        
```

File: exercises/concept/secrets/secrets_test.py
Comment (BethanyG) [inline exercises/concept/secrets/secrets_test.py:9]: All of our test cases should have a human-friendly `msg` parameter for return in the UI.  Please see [this example](https://github.com/exercism/python/blob/main/exercises/concept/guidos-gorgeous-lasagna/lasagna_test.py) from the Guidos Gorgeous Lasagna exercise.

File: config.json
Comment (BethanyG) [inline config.json:2283]: This exercise is in the incorrect section of the `config.json`, and is failing CI.  It needs to be moved to the `concept exercise` section.

File: exercises/concept/secrets/.meta/config.json
Comment (BethanyG) [inline exercises/concept/secrets/.meta/config.json:3]:
```suggestion
  "authors": ["Steffan153"],
  "contributors": ["bethanyg"]
```

Comment (BethanyG) [reply exercises/concept/secrets/.meta/config.json:3]: ..you are using my comments, after all.  😉

File: config.json
Comment (Steffan153) [reply config.json:2283]: No wonder it wasn't matching up with the others. Lol

**Comment by Steffan153** [comment]
Btw, how do I tell if the CI is failing or not? :thinking:

**Comment by BethanyG** [comment]
> Btw, how do I tell if the CI is failing or not? 🤔

It usually shows here on the main page...buuut the workflow needs to be approved by a maintainer, so if you then do another push, it gets hidden until the workflow gets approved/run again.

**Comment by BethanyG** [comment]
_ooof_  I think I need to take a break and look at this with fresh eyes in the morning (_after 6am Pacific_).  Apologies, @Steffan153  -- I am a bit blurry at the moment, so please bear with me.

**Comment by Steffan153** [comment]
I'm just 1 hour ahead of you and am a super heavy sleeper (I'm about to buy one of those bed shakers), so that should be plenty good :)

**Comment by Steffan153** [comment]
Hey, just wondering what happened with this?

**Comment by BethanyG** [comment]
Hi @Steffan153 -- I got pulled into troubleshooting, and then this slipped through the cracks.  But I am reviewing today.  Apologies.

**Comment by BethanyG** [comment]
@Steffan153  -- I have really been struggling with this exercise, and how to adapt it better to Python.
Much of it feels like an exercise on `higher order functions` as opposed to `lambda` itself.  Especially the line from the instructions:  

> In this exercise, you've been tasked with writing the software for an encryption device that works by performing transformations on data. You need a way to flexibly create complicated functions by combining simpler functions together.

The most "natural" way to do this in Python would be to simply write a function that takes or returns a function -- not by writing/returning a `lambda`.

I am also struggling with this, from the introduction:

> Python has a [`lambda`][lambdas] expression, which is a shorter way to create functions than using `def`.

...because it is perfectly fine in Python to define a one-line named function.  Mind you, it is considered pretty bad form, but this will compile and run without issue:  `def one_liner(first, second): first=first**2; second=second-3; return first + second`


So I am trying to think through how to modify tasks 1 & 2 to better show where or _why_ you would need a `lambda` _instead_ of a simple (and named) one-line function.

I think task 3 can be modified to use `max()` and or `min()` with a `key function`.  

Task 5 ... _might_ be able to be modified to pull functions from a `dict` based on a `key function`.  The other thought is to set up a scenario where one or more of these `functions` are callbacks that need parameters ... which would be where you might use a `lambda` instead.

**The long and short :**  This exercise needs work - but it is around clarifying what topics are, and are not covered.  I would suggest in-line edits, but I am not sure that is useful, since they would be less edits, and more entire re-writes.  So I am  inclined to "optimistically merge" this, and then discuss how we re-work it.

The other option is to turn this exercise into the start of a concept exercise on `higher order functions`, and pick a different scenario/topic for `lambda` that is more tailored to the restrictions Python places on them.  I came across [this](https://github.com/dabeaz-course/practical-python/blob/master/Notes/07_Advanced_Topics/02_Anonymous_function.md) from David Beazley, which could maybe be inspiration for a different exercise on `lambda`.

Thoughts on how you would like to proceed?  Let me know.  Thanks!

**Comment by Steffan153** [comment]
What if we have this learn both the Lambdas concept and the Higher-Order Functions? Sounds like they would go well together.

**Comment by BethanyG** [comment]
@Steffan153  -- are you still up for working on this?  Let me know...thanks!

**Comment by BethanyG** [comment]
@Steffan153  -- Just checking in again.  Are you still working on this?

**Comment by Steffan153** [comment]
Yeah I'll get to it sometime...

**Comment by BethanyG** [comment]
I'm going to close for now to clear the queue.  Ping @J08K  or myself when you are ready, and we will happily re-open.

**Comment by safwansamsudeen** [comment]
@Steffan153 can we work on this together, if you're fine with that? This seems interesting.

**Comment by BethanyG** [comment]
@safwansamsudeen  -- This exercise is closed to contribution at this time.

**Pull request closed without merging**
